### ISSUE TITLE
iperf3 reporting excessive retransmits 

### ISSUE THREAD

**Issue by yghorbal** [esnet/iperf#833]
# Context
Trying to track packet loss on the network. I'm running iperf3 on FreeBSD based host acting as iperf3 server and a CentOS 7.6 host acting as iperf3 client. I'm using these options:
on server: 
```
./iperf3 -s -p 55556
``` 
on client:
```
./iperf3 --cport 3333 -t 60 -Rc 192.168.133.2 -p 55556
```
I'm using the `-R` to be in same situation of my applications where client connects to server, requests data and data is transferred from server to client. I'm using fancy port numbers to make tcpdump filters predictable.

Server side is reporting rather huge packet loss:
```
-----------------------------------------------------------
Server listening on 55556
-----------------------------------------------------------
Accepted connection from 192.168.131.140, port 60244
[  5] local 192.168.133.2 port 55556 connected to 192.168.131.140 port 3333
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  1.00 GBytes  8.60 Gbits/sec    0    472 KBytes
[  5]   1.00-2.00   sec  1.09 GBytes  9.39 Gbits/sec    0    676 KBytes
[  5]   2.00-3.00   sec  1.09 GBytes  9.39 Gbits/sec    0    820 KBytes
[  5]   3.00-4.00   sec  1.09 GBytes  9.39 Gbits/sec    0    954 KBytes
[  5]   4.00-5.00   sec  1.04 GBytes  8.98 Gbits/sec    7    107 KBytes
[  5]   5.00-6.00   sec   870 MBytes  7.30 Gbits/sec   24    182 KBytes
[  5]   6.00-7.00   sec   831 MBytes  6.97 Gbits/sec   45   84.8 KBytes
[  5]   7.00-8.00   sec   796 MBytes  6.67 Gbits/sec   43   94.7 KBytes
[  5]   8.00-9.00   sec  1.05 GBytes  9.02 Gbits/sec    9    115 KBytes
[  5]   9.00-10.00  sec   913 MBytes  7.66 Gbits/sec   40   83.4 KBytes
[  5]   9.00-10.00  sec   913 MBytes  7.66 Gbits/sec   40   83.4 KBytes
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  9.96 GBytes  8.55 Gbits/sec  168             sender
iperf3: the client has terminated
```

client side pcap can be downloaded [here](https://ln.sync.com/dl/44f5d34b0/view/default/7852894110005#h8bb5gda-frp78vyt-yapws5gx-kidhq4mb)
server side pcap can be downloaded [here](https://cp.sync.com/dl/323677c10#gh3gy88t-ge3gyyxn-hka5ya6q-q6ydp5fn)

pcaps on both sides show that somehow the client is receiving packets out of order or sometimes missing segments, to which it reacts with selective ACKs. As far I can tell this is "normal" behaviour and clearly shows some issues on the path between those two hosts. 
However, what I can't explain is the huge amount a TCP DUP ACKs sent by the server for no obvious reason. In this setup client is connecting to server and due to `-R` option, data is flowing from server to client. The only data I see coming from the client is at test initialisation where client sends a small data packet, that gets ACKed quickly afterword.

**Is there any known scenario where iperf3 at server side can trigger this DUP ACK sent repeatedly?** 

* Version of iperf3:
iperf 3.4 (cJSON 1.5.2)
Optional features available: CPU affinity setting, IPv6 flow label, TCP congestion algorithm setting, sendfile / zerocopy, authentication
(same version on client and server)

* Hardware:
Standard x86 servers with 8 cores and 64G RAM.
Network is 10Gb/s Intel NICs.
Network gears between test boxes are Arista
FreeBSD host is connected through a lagg on top of two 10Gb/s NICs.

* Operating system (and distribution, if any):
server: FreeBSD based
client: CentOS 7.6

**Comment by bmah888** [comment]
This is an interesting problem, but I don't see anything there that points to there being an iperf3 issue.

**Comment by bmah888** [comment]
Closing with this pointer as a possible starting point for future (outside of iperf3 scope) investigation:

https://meetings.internet2.edu/2019-technology-exchange/detail/10005523/

**Comment by davidBar-On** [comment]
Just FYI.  There doesn't seem to be almost any dup acks.  Acks sent by TCP are not counted in the sequence number.  Therefore, the receiver side sequence number remains 38 (after the 37 data bytes sent in line 4).  This is why sequence 38 is seen in all acks: receiver acks are with this sequence, and sender packets always piggyback the ack for last sequence received.

There is however an issue, as it seems that the reverse option did not take effect, and the client is the sender.  That may be because -R comes before -c.  If there was such problem in the code, it seems that it was fixed since version 3.4.

**Comment by yghorbal** [comment]
@davidBar-On 

 Thank you for taking the time to look into this.
 From the provided PCAPs I can clearly see that the server is actually sending data. For instance 
 Server is 192.168.133.2 and client is 192.168.131.140.
 As for acks, I can clearly see from wireshark that there are actually many DUP acks. In client PCAP, Frames 383048 -> 383080 are an exemple (we can see equivalent of this sequence in server PCAP too) Theses are only TCP ACK packets (with 0 data) sent by the server and not Piggy backed with data packets.
 I still think the issue is real (or at least it's not a normal behaviour)

 @bmah888 thank you for the link!

Youssef

**Comment by davidBar-On** [comment]
@yghorbal
I now see that the server is the sender and that the -R works fine.  Sorry for the mistake (I did have the right server IP based on the iperf -c command you listed).

Regarding the issue.  It doesn't seem to me that there are many dup acks.  With TCP it is expected that there may be some temporary delays in the network that cause dup acks from time to time.

Regarding the example you gave. I noticed that at that point of time there is almost exactly 1MB difference between the sent sequence and the received sequence in the ack.  E.g. in the server's log line 305982 the sequence sent is 1341486985, while in line 305983 ack is received for 1340453113.  This is ‭1,033,872‬ bytes difference which is almost exactly 1MB (‭1,048,576‬ bytes).  It seems (just a guess) that the congestion control algorithm used is configured to stop sending data when the difference is 1MB, until almost all data that was already sent is acked.  In addition it seems that it is designed/configured send zero length packets during that time.

If you know the congestion control algorithm used, it may help to evaluate this behavior.  E.g., it may be that the algorithm is designed for networks were for some reason the sender must continuously send something.
